include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+                                  const Function<spacedim, typename VectorType::value_type> &function,
+                                  VectorType                                                &vec_result,
+                                  const bool                                                 enforce_zero_boundary,
+                                  const Quadrature<dim-1>                                   &q_boundary,
+                                  const bool                                                 project_to_boundary_first)
+    {
+      switch (dof.get_fe().degree)
+        {
+        case 1:
+          dealii::VectorTools::project_generic<components, 1>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 2:
+          dealii::VectorTools::project_generic<components, 2>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 3:
+          dealii::VectorTools::project_generic<components, 3>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 4:
+          dealii::VectorTools::project_generic<components, 4>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 5:
+          dealii::VectorTools::project_generic<components, 5>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 6:
+          dealii::VectorTools::project_generic<components, 6>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 7:
+          dealii::VectorTools::project_generic<components, 7>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 8:
+          dealii::VectorTools::project_generic<components, 8>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
+        }
+    }
+
+
+
+    // Helper interface for the matrix-free implementation of project().
+    // Used to determine the number of components.
+    template <int dim, typename VectorType, int spacedim>
+    void project_parallel (const Mapping<dim, spacedim>           &mapping,
```

Comment:
i would also rename this one to something like `project_matrix_free_component`, again to better indicate that you switch based on the component.